File: sdks/elixir/lib/kyuubiki_sdk/workflow_builders.ex
defmodule KyuubikiSdk.WorkflowBuilders do
  @moduledoc "Builder helpers for workflow dataset and graph contracts."

  alias KyuubikiSdk.WorkflowContracts

  def schema_ref(schema, version) do
    %{"schema" => schema, "version" => version}
  end

  def axis(axis_id, attrs \\ %{}) do
    %{"id" => axis_id}
    |> maybe_put("label", attr(attrs, :label))
    |> maybe_put("size", attr(attrs, :size))
    |> maybe_put("semantic", attr(attrs, :semantic))
  end

  def shape(attrs \\ %{}) do
    %{}
    |> maybe_put("axes", attr(attrs, :axes))
  end

  def dataset_value(value_id, data_class, element_type, attrs \\ %{}) do
    %{
      "id" => value_id,
      "data_class" => data_class,
      "element_type" => element_type,
      "shape" => attr(attrs, :shape) || %{}
    }
    |> maybe_put("semantic_type", attr(attrs, :semantic_type))
    |> maybe_put("unit", attr(attrs, :unit))
    |> maybe_put("encoding", attr(attrs, :encoding))
    |> maybe_put("schema_ref", attr(attrs, :schema_ref))
  end

  def dataset_contract(contract_id, version, values, attrs \\ %{}) do
    contract =
      %{
        "schema_version" => WorkflowContracts.workflow_dataset_schema_version(),
        "id" => contract_id,
        "version" => version,
        "values" => values
      }
      |> maybe_put("name", attr(attrs, :name))
      |> maybe_put("description", attr(attrs, :description))
      |> maybe_put("metadata", attr(attrs, :metadata))

    if Keyword.get(List.wrap(attrs), :validate, true) == false or Map.get(attrs, :validate) == false do
      contract
    else
      case WorkflowContracts.validate_dataset_contract(contract) do
        {:ok, validated} -> validated
        {:error, error} -> raise error
      end
    end
  end

  def port(port_id, artifact_type, attrs \\ %{}) do
    %{"id" => port_id, "artifact_type" => artifact_type}
    |> maybe_put("name", attr(attrs, :name))
    |> maybe_put("required", attr(attrs, :required))
    |> maybe_put("cardinality", attr(attrs, :cardinality))
    |> maybe_put("dataset_value", attr(attrs, :dataset_value))
  end

  def defaults(attrs \\ %{}) do
    %{}
    |> maybe_put("cache_policy", attr(attrs, :cache_policy))
    |> maybe_put("orchestrated", attr(attrs, :orchestrated))
    |> maybe_put("dispatch_policy", attr(attrs, :dispatch_policy))
    |> maybe_put("placement_tags", attr(attrs, :placement_tags))
    |> maybe_put("required_capabilities", attr(attrs, :required_capabilities))
  end

  def operator_fetch_entry(node_id, operator_id, attrs \\ %{}) do
    %{"node_id" => node_id, "operator_id" => operator_id}
    |> maybe_put("package_ref", attr(attrs, :package_ref))
    |> maybe_put("version", attr(attrs, :version))
    |> maybe_put("integrity", attr(attrs, :integrity))
    |> maybe_put("cache_scope", attr(attrs, :cache_scope))
  end

  def node(node_id, kind, attrs \\ %{}) do
    %{
      "id" => node_id,
      "kind" => kind,
      "inputs" => attr(attrs, :inputs) || [],
      "outputs" => attr(attrs, :outputs) || []
    }
    |> maybe_put("operator_id", attr(attrs, :operator_id))
    |> maybe_put("name", attr(attrs, :name))
    |> maybe_put("description", attr(attrs, :description))
    |> maybe_put("config", attr(attrs, :config))
    |> maybe_put("cache_policy", attr(attrs, :cache_policy))
    |> maybe_put("placement_tags", attr(attrs, :placement_tags))
    |> maybe_put("required_capabilities", attr(attrs, :required_capabilities))
  end

  def edge(edge_id, from_node, from_port, to_node, to_port, artifact_type, attrs \\ %{}) do
    %{
      "id" => edge_id,
      "from" => %{"node" => from_node, "port" => from_port},
      "to" => %{"node" => to_node, "port" => to_port},
      "artifact_type" => artifact_type
    }
    |> maybe_put("dataset_value", attr(attrs, :dataset_value))
  end

  def graph(graph_id, name, version, entry_nodes, nodes, edges, attrs \\ %{}) do
    graph =
      %{
        "schema_version" => WorkflowContracts.workflow_graph_schema_version(),
        "id" => graph_id,
        "name" => name,
        "version" => version,
        "entry_nodes" => entry_nodes,
        "nodes" => nodes,
        "edges" => edges
      }
      |> maybe_put("description", attr(attrs, :description))
      |> maybe_put("dataset_contract", attr(attrs, :dataset_contract))
      |> maybe_put("output_nodes", attr(attrs, :output_nodes))
      |> maybe_put("defaults", attr(attrs, :defaults))
      |> maybe_put("dispatch_policy", attr(attrs, :dispatch_policy))
      |> maybe_put("operator_fetch_plan", attr(attrs, :operator_fetch_plan))
      |> maybe_put("placement_tags", attr(attrs, :placement_tags))
      |> maybe_put("required_capabilities", attr(attrs, :required_capabilities))

    if Keyword.get(List.wrap(attrs), :validate, true) == false or Map.get(attrs, :validate) == false do
      graph
    else
      case WorkflowContracts.validate_graph(graph) do
        {:ok, validated} -> validated
        {:error, error} -> raise error
      end
    end
  end

  defp maybe_put(map, _key, nil), do: map
  defp maybe_put(map, key, value), do: Map.put(map, key, value)
  defp attr(attrs, key), do: Map.get(attrs, key, Map.get(attrs, Atom.to_string(key)))
end

File: sdks/python/kyuubiki_sdk/workflow_builders.py
from __future__ import annotations

from typing import Any

from .workflow_contracts import (
    WORKFLOW_DATASET_SCHEMA_VERSION,
    WORKFLOW_GRAPH_SCHEMA_VERSION,
    validate_workflow_dataset_contract,
    validate_workflow_graph,
)


def build_workflow_schema_ref(schema: str, version: str) -> dict[str, str]:
    return {"schema": schema, "version": version}


def build_workflow_axis(
    axis_id: str,
    *,
    label: str | None = None,
    size: int | None = None,
    semantic: str | None = None,
) -> dict[str, Any]:
    axis: dict[str, Any] = {"id": axis_id}
    if label is not None:
        axis["label"] = label
    if size is not None:
        axis["size"] = size
    if semantic is not None:
        axis["semantic"] = semantic
    return axis


def build_workflow_shape(*, axes: list[dict[str, Any]] | None = None) -> dict[str, Any]:
    shape: dict[str, Any] = {}
    if axes is not None:
        shape["axes"] = axes
    return shape


def build_workflow_dataset_value(
    value_id: str,
    *,
    data_class: str,
    element_type: str,
    shape: dict[str, Any] | None = None,
    semantic_type: str | None = None,
    unit: str | None = None,
    encoding: str | None = None,
    schema_ref: dict[str, Any] | None = None,
) -> dict[str, Any]:
    value: dict[str, Any] = {
        "id": value_id,
        "data_class": data_class,
        "element_type": element_type,
        "shape": shape or {},
    }
    if semantic_type is not None:
        value["semantic_type"] = semantic_type
    if unit is not None:
        value["unit"] = unit
    if encoding is not None:
        value["encoding"] = encoding
    if schema_ref is not None:
        value["schema_ref"] = schema_ref
    return value


def build_workflow_dataset_contract(
    contract_id: str,
    *,
    version: str,
    values: list[dict[str, Any]],
    name: str | None = None,
    description: str | None = None,
    metadata: dict[str, str] | None = None,
    validate: bool = True,
) -> dict[str, Any]:
    contract: dict[str, Any] = {
        "schema_version": WORKFLOW_DATASET_SCHEMA_VERSION,
        "id": contract_id,
        "version": version,
        "values": values,
    }
    if name is not None:
        contract["name"] = name
    if description is not None:
        contract["description"] = description
    if metadata is not None:
        contract["metadata"] = metadata
    return validate_workflow_dataset_contract(contract) if validate else contract


def build_workflow_port(
    port_id: str,
    *,
    artifact_type: str,
    name: str | None = None,
    required: bool | None = None,
    cardinality: str | None = None,
    dataset_value: str | None = None,
) -> dict[str, Any]:
    port: dict[str, Any] = {"id": port_id, "artifact_type": artifact_type}
    if name is not None:
        port["name"] = name
    if required is not None:
        port["required"] = required
    if cardinality is not None:
        port["cardinality"] = cardinality
    if dataset_value is not None:
        port["dataset_value"] = dataset_value
    return port


def build_workflow_node(
    node_id: str,
    *,
    kind: str,
    inputs: list[dict[str, Any]],
    outputs: list[dict[str, Any]],
    operator_id: str | None = None,
    name: str | None = None,
    description: str | None = None,
    config: dict[str, Any] | None = None,
    cache_policy: str | None = None,
) -> dict[str, Any]:
    node: dict[str, Any] = {
        "id": node_id,
        "kind": kind,
        "inputs": inputs,
        "outputs": outputs,
    }
    if operator_id is not None:
        node["operator_id"] = operator_id
    if name is not None:
        node["name"] = name
    if description is not None:
        node["description"] = description
    if config is not None:
        node["config"] = config
    if cache_policy is not None:
        node["cache_policy"] = cache_policy
    return node


def build_workflow_edge(
    edge_id: str,
    *,
    from_node: str,
    from_port: str,
    to_node: str,
    to_port: str,
    artifact_type: str,
    dataset_value: str | None = None,
) -> dict[str, Any]:
    edge: dict[str, Any] = {
        "id": edge_id,
        "from": {"node": from_node, "port": from_port},
        "to": {"node": to_node, "port": to_port},
        "artifact_type": artifact_type,
    }
    if dataset_value is not None:
        edge["dataset_value"] = dataset_value
    return edge


def build_workflow_graph(
    graph_id: str,
    *,
    name: str,
    version: str,
    entry_nodes: list[str],
    nodes: list[dict[str, Any]],
    edges: list[dict[str, Any]],
    description: str | None = None,
    dataset_contract: dict[str, Any] | None = None,
    output_nodes: list[str] | None = None,
    defaults: dict[str, Any] | None = None,
    validate: bool = True,
) -> dict[str, Any]:
    graph: dict[str, Any] = {
        "schema_version": WORKFLOW_GRAPH_SCHEMA_VERSION,
        "id": graph_id,
        "name": name,
        "version": version,
        "entry_nodes": entry_nodes,
        "nodes": nodes,
        "edges": edges,
    }
    if description is not None:
        graph["description"] = description
    if dataset_contract is not None:
        graph["dataset_contract"] = dataset_contract
    if output_nodes is not None:
        graph["output_nodes"] = output_nodes
    if defaults is not None:
        graph["defaults"] = defaults
    return validate_workflow_graph(graph) if validate else graph
